File: aoc-2025/cpp/src/day07.cpp
#include <algorithm>
#include <cstddef>
#include <fstream>
#include <iostream>
#include <string>
#include <vector>

const std::string DEFAULT_INPUT = "";
const std::string DEFAULT_OUTPUT = "";

struct GridInfo {
  std::vector<std::string> grid;
  int st_r = -1;
  int st_c = -1;
  int ht = 0;
  int wd = 0;
};

GridInfo load_grid(const std::string &filename) {
  GridInfo info;
  std::ifstream file(filename);

  std::string line;
  while (std::getline(file, line)) {
    if (!line.empty() && line.back() == '\r')
      line.pop_back();
    if (!line.empty())
      info.grid.push_back(line);
  }

  if (info.grid.empty())
    return info;

  info.ht = (info.grid.size());
  info.wd = (info.grid[0].size());

  for (int r = 0; r < info.ht; r++) {
    size_t pos = info.grid[r].find('S');
    if (pos != std::string::npos) {
      info.st_r = r;
      info.st_c = pos;
      break;
    }
  }
  return info;
}

long long part_one(const std::string &filename) {
  auto info = load_grid(filename);
  if (info.grid.empty() || info.st_r == -1)
    return 0;

  std::vector<int> curr;
  curr.push_back(info.st_c);

  long long ans = 0;

  for (int r = info.st_r; r < info.ht; r++) {
    if (curr.empty())
      break;

    std::vector<int> next;
    next.reserve(curr.size() * 2);

    for (int c : curr) {
      if (c < 0 || c >= info.wd)
        continue;
      char cell = info.grid[r][c];
      if (cell == '^') {
        ans++;
        if (c - 1 >= 0)
          next.push_back(c - 1);
        if (c + 1 < info.wd)
          next.push_back(c + 1);
      } else {
        next.push_back(c);
      }
    }

    std::sort(next.begin(), next.end());
    next.erase(std::unique(next.begin(), next.end()), next.end());
    curr = next;
  }

  return ans;
}

long long part_two(const std::string &filename) {
  auto info = load_grid(filename);
  if (info.grid.empty() || info.st_r == -1)
    return 0;

  std::vector<long long> cnt(info.wd, 0);
  cnt[info.st_c] = 1;
  for (int r = info.st_r; r < info.ht; r++) {
    std::vector<long long> next(info.wd, 0);

    for (int c = 0; c < info.wd; c++) {
      if (cnt[c] == 0)
        continue;
      char cell = info.grid[r][c];

      if (cell == '^') {
        if (c - 1 >= 0) {
          next[c - 1] += cnt[c];
        }
        if (c + 1 < info.wd) {
          next[c + 1] += cnt[c];
        }
      } else {
        next[c] += cnt[c];
      }
    }
    cnt = next;
  }

  long long ans = 0;
  for (long long count : cnt)
    ans += count;
  return ans;
}

int main(int argc, char *argv[]) {
  std::string input_file = (argc > 1) ? argv[1] : DEFAULT_INPUT;
  std::cout << "Reading from: " << input_file << std::endl;

  auto p1 = part_one(input_file);
  auto p2 = part_two(input_file);

  std::cout << "Part 1: " << p1 << std::endl;
  std::cout << "Part 2: " << p2 << std::endl;

  std::ofstream oFile(DEFAULT_OUTPUT);
  if (oFile.is_open())
    oFile << p1 << "\n" << p2 << "\n";

  return 0;
}

File: aoc-2025/rust/src/bin/day07.rs
//! https://adventofcode.com/2025/day/7
use std::env;
use std::fs;
use std::path::Path;

const DEFAULT_INPUT: &str = "";
const DEFAULT_OUTPUT: &str = "";

fn part_one(content: &str) -> i32 {
    let grid: Vec<Vec<char>> = content
        .lines()
        .filter(|l| !l.is_empty())
        .map(|l| l.chars().collect())
        .collect();

    if grid.is_empty() {
        return 0;
    }
    let ht = grid.len();
    let wd = grid[0].len();
    let mut st = None;

    for r in 0..ht {
        for c in 0..wd {
            if grid[r][c] == 'S' {
                st = Some((r, c));
                break;
            }
        }
        if st.is_some() {
            break;
        }
    }

    let (st_r, st_c) = match st {
        Some(p) => p,
        None => return 0,
    };

    let mut curr = vec![st_c];
    let mut ans = 0;
    for r in st_r..ht {
        if curr.is_empty() {
            break;
        }
        let mut next = Vec::new();
        for c in curr {
            if c >= wd {
                continue;
            }

            let cell = grid[r][c];

            if cell == '^' {
                ans += 1;
                if c > 0 {
                    next.push(c - 1);
                }
                if c + 1 < wd {
                    next.push(c + 1);
                }
            } else {
                next.push(c);
            }
        }
        next.sort_unstable();
        next.dedup();
        curr = next;
    }
    ans
}

fn part_two(content: &str) -> i64 {
    let grid: Vec<Vec<char>> = content
        .lines()
        .filter(|l| !l.is_empty())
        .map(|l| l.chars().collect())
        .collect();

    if grid.is_empty() {
        return 0;
    }
    let ht = grid.len();
    let wd = grid[0].len();
    let mut st = None;

    for r in 0..ht {
        for c in 0..wd {
            if grid[r][c] == 'S' {
                st = Some((r, c));
                break;
            }
        }
        if st.is_some() {
            break;
        }
    }

    let (st_r, st_c) = match st {
        Some(p) => p,
        None => return 0,
    };

    let mut cnt = vec![0i64; wd];
    cnt[st_c] = 1;

    for r in st_r..ht {
        let mut next = vec![0i64; wd];
        for c in 0..wd {
            if cnt[c] == 0 {
                continue;
            }

            let cell = grid[r][c];

            if cell == '^' {
                if c > 0 {
                    next[c - 1] += cnt[c];
                }
                if c + 1 < wd {
                    next[c + 1] += cnt[c];
                }
            } else {
                next[c] += cnt[c];
            }
        }
        cnt = next;
    }
    cnt.iter().sum()
}

fn main() {
    let args: Vec<String> = env::args().collect();
    let input_path = args.get(1).map(|s| s.as_str()).unwrap_or(DEFAULT_INPUT);
    println!("Reading from: {}", input_path);

    let content = fs::read_to_string(input_path).expect("Could not read input file");

    let p1 = part_one(&content);
    let p2 = part_two(&content);

    println!("Part 1: {}", p1);
    println!("Part 2: {}", p2);

    if let Some(parent) = Path::new(DEFAULT_OUTPUT).parent() {
        let _ = fs::create_dir_all(parent);
    }
    let _ = fs::write(DEFAULT_OUTPUT, format!("{}\n{}\n", p1, p2));
}
